# EDA

> fuente: Air bnd New York

In [20]:
import numpy as numpy
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

### Empezamos leyendo con Pandas el archivo csv

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/data-preprocessing-project-tutorial/main/AB_NYC_2019.csv")
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

1. train_date.describe().
- Este método genera estadísticas descriptivas que resumen la tendencia central, dispersión y forma de la distribución de cada columna 
- Para columnas numéricas muestra:
  count, mean (media), std (desviación estándar), min, percentiles (25%, 50% = mediana, 75%), y max 
- Para columnas de tipo “objeto” (strings, fechas…), usando include='all', muestra:
count, unique, top (valor más frecuente) y freq (frecuencia) 
2. El sufijo .T
T es un accesorio para .transpose(), que rota la tabla intercambiando filas por columnas

In [3]:
train_date = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/data-preprocessing-project-tutorial/main/AB_NYC_2019.csv")
train_date.describe().T

count          mean           std  \
id                              48895.0  1.901714e+07  1.098311e+07   
host_id                         48895.0  6.762001e+07  7.861097e+07   
latitude                        48895.0  4.072895e+01  5.453008e-02   
longitude                       48895.0 -7.395217e+01  4.615674e-02   
price                           48895.0  1.527207e+02  2.401542e+02   
minimum_nights                  48895.0  7.029962e+00  2.051055e+01   
number_of_reviews               48895.0  2.327447e+01  4.455058e+01   
reviews_per_month               38843.0  1.373221e+00  1.680442e+00   
calculated_host_listings_count  48895.0  7.143982e+00  3.295252e+01   
availability_365                48895.0  1.127813e+02  1.316223e+02   

                                       min           25%           50%  \
id                              2539.00000  9.471945e+06  1.967728e+07   
host_id                         2438.00000  7.822033e+06  3.079382e+07   
latitude                          40.49979  4.069010e+01  4.072307e+01   
longitude                        -74.24442 -7.398307e+01 -7.395568e+01   
price                              0.00000  6.900000e+01  1.060000e+02   
minimum_nights                     1.00000  1.000000e+00  3.000000e+00   
number_of_reviews                  0.00000  1.000000e+00  5.000000e+00   
reviews_per_month                  0.01000  1.900000e-01  7.200000e-01   
calculated_host_listings_count     1.00000  1.000000e+00  1.000000e+00   
availability_365                   0.00000  0.000000e+00  4.500000e+01   

                                         75%           max  
id                              2.915218e+07  3.648724e+07  
host_id                         1.074344e+08  2.743213e+08  
latitude                        4.076311e+01  4.091306e+01  
longitude                      -7.393627e+01 -7.371299e+01  
price                           1.750000e+02  1.000000e+04  
minimum_nights                  5.000000e+00  1.250000e+03  
number_of_reviews               2.400000e+01  6.290000e+02  
reviews_per_month               2.020000e+00  5.850000e+01  
calculated_host_listings_count  2.000000e+00  3.270000e+02  
availability_365                2.270000e+02  3.650000e+02

- False significa que esa fila no es un duplicado.
- True indicaría que esa fila sí es un duplicado de una fila previa.

In [4]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
48890    False
48891    False
48892    False
48893    False
48894    False
Length: 48895, dtype: bool

In [5]:
df[df.duplicated()]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

Obtendremos las dimensiones: vemos que tenemos 48895 filas y 16 columnas

In [6]:
df.shape

(48895, 16)

> Aqui vemos que la columna name tiene mas de un valor nulo, ya que nos devuelve True (16)
> Esto tambien nos muestra las filas donde esa columna esta vacia, por lo que procedemos a eliminarla.

In [15]:
df['name'].isnull().sum()
df[df['name'].isnull()]

id name   host_id host_name neighbourhood_group  \
2854    1615764  NaN   6676776     Peter           Manhattan   
3703    2232600  NaN  11395220      Anna           Manhattan   
5775    4209595  NaN  20700823     Jesse           Manhattan   
5975    4370230  NaN  22686810   Michaël           Manhattan   
6269    4581788  NaN  21600904     Lucie            Brooklyn   
6567    4756856  NaN   1832442  Carolina            Brooklyn   
6605    4774658  NaN  24625694      Josh           Manhattan   
8841    6782407  NaN  31147528  Huei-Yin            Brooklyn   
11963   9325951  NaN  33377685  Jonathan           Manhattan   
12824   9787590  NaN  50448556    Miguel           Manhattan   
13059   9885866  NaN  37306329  Juliette           Manhattan   
13401  10052289  NaN  49522403   Vanessa            Brooklyn   
15819  12797684  NaN  69715276       Yan           Manhattan   
16071  12988898  NaN  71552588    Andrea               Bronx   
18047  14135050  NaN  85288337      Jeff            Brooklyn   
28889  22275821  NaN  49662398  Kathleen            Brooklyn   

            neighbourhood  latitude  longitude        room_type  price  \
2854    Battery Park City  40.71239  -74.01620  Entire home/apt    400   
3703         East Village  40.73215  -73.98821  Entire home/apt    200   
5775    Greenwich Village  40.73473  -73.99244  Entire home/apt    225   
5975               Nolita  40.72046  -73.99550  Entire home/apt    215   
6269         Williamsburg  40.71370  -73.94378     Private room    150   
6567             Bushwick  40.70046  -73.92825     Private room     70   
6605   Washington Heights  40.85198  -73.93108     Private room     40   
8841         Williamsburg  40.71354  -73.93882     Private room     45   
11963      Hell's Kitchen  40.76436  -73.98573  Entire home/apt    190   
12824              Harlem  40.80316  -73.95189  Entire home/apt    300   
13059           Chinatown  40.71632  -73.99328     Private room     67   
13401         Brownsville  40.66409  -73.92314     Private room     50   
15819     Upper West Side  40.79843  -73.96404     Private room    100   
16071             Fordham  40.86032  -73.88493      Shared room    130   
18047  Bedford-Stuyvesant  40.69421  -73.93234     Private room     70   
28889            Bushwick  40.69546  -73.92741  Entire home/apt    110   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
2854             1000                  0         NaN                NaN   
3703                1                 28  2015-06-08               0.45   
5775                1                  1  2015-01-01               0.02   
5975                7                  5  2016-01-02               0.09   
6269                1                  0         NaN                NaN   
6567                1                  0         NaN                NaN   
6605                1                  0         NaN                NaN   
8841                1                  0         NaN                NaN   
11963               4                  1  2016-01-05               0.02   
12824               5                  0         NaN                NaN   
13059               4                  0         NaN                NaN   
13401               3                  3  2016-08-18               0.07   
15819               1                  0         NaN                NaN   
16071               1                  0         NaN                NaN   
18047               3                  0         NaN                NaN   
28889               4                  5  2018-08-13               0.27   

       calculated_host_listings_count  availability_365  
2854                                1               362  
3703                                1               341  
5775                                1                 0  
5975                                1                 0  
6269                                1                 0  
6567                                1    

> Aqui tambien vemos todas las columnas con mas de un valor nulo

In [19]:
df.isnull().sum()[df.isnull().sum() > 0]

name                    16
host_name               21
last_review          10052
reviews_per_month    10052
dtype: int64

> Decido aliminar algunas columnas con demasiados valores nulos.
> Elimino last_review, reviews_per:months, host_name y name

In [30]:
df = df.drop("name")

KeyError: "['name'] not found in axis"

Clase: Este objeto es un pandas.DataFrame, una estructura tabular como una hoja de Excel.

RangeIndex: El índice va del 0 al 48894, es decir, hay 48,895 filas (registros de alojamiento).

Columnas: Hay 16 columnas (características de cada alojamiento).

| #  | Columna                          | Non-Null Count | Tipo de dato | Significado                                       |
| -- | -------------------------------- | -------------- | ------------ | ------------------------------------------------- |
| 0  | `id`                             | 48895          | `int64`      | Identificador único del anuncio.                  |
| 1  | `name`                           | 48879          | `object`     | Nombre o título del anuncio. Hay 16 nulos.        |
| 2  | `host_id`                        | 48895          | `int64`      | ID único del anfitrión.                           |
| 3  | `host_name`                      | 48874          | `object`     | Nombre del anfitrión. Hay 21 nulos.               |
| 4  | `neighbourhood_group`            | 48895          | `object`     | Grupo o distrito general (ej. Manhattan).         |
| 5  | `neighbourhood`                  | 48895          | `object`     | Vecindario específico.                            |
| 6  | `latitude`                       | 48895          | `float64`    | Coordenada de latitud.                            |
| 7  | `longitude`                      | 48895          | `float64`    | Coordenada de longitud.                           |
| 8  | `room_type`                      | 48895          | `object`     | Tipo de habitación (entero, compartido, etc).     |
| 9  | `price`                          | 48895          | `int64`      | Precio por noche.                                 |
| 10 | `minimum_nights`                 | 48895          | `int64`      | Número mínimo de noches por reserva.              |
| 11 | `number_of_reviews`              | 48895          | `int64`      | Cantidad de reseñas del anuncio.                  |
| 12 | `last_review`                    | 38843          | `object`     | Fecha de la última reseña. Hay muchos nulos.      |
| 13 | `reviews_per_month`              | 38843          | `float64`    | Reseñas promedio por mes. También tiene nulos.    |
| 14 | `calculated_host_listings_count` | 48895          | `int64`      | Cantidad de anuncios activos del anfitrión.       |
| 15 | `availability_365`               | 48895          | `int64`      | Disponibilidad del alojamiento en el año (0–365). |
